File: code/python/05.1.1_ColorfulLED/ColorfulLED.py
#!/usr/bin/env python3
########################################################################################
# Filename    : ColorfulLED.py
# Description : Random color change ColorfulLED
# Author      : www.freenove.com
# modification: 2023/05/11
########################################################################################

########################################################################################
##
## import `RGBLED` functionality from `gpiozero` library
##
from gpiozero import RGBLED

########################################################################################
##
## import the `time` and `random` libs
##
import time
import random


########################################################################################
##
## instantiating the RGBLED class with specified pins and type for controlling the LED
##
## use first option for a common anode RGBLED and second for a common cathode RGBLED
##
## define the pins for R:11,G:12,B:13 
##
#led = RGBLED(red="J8:11", green="J8:12", blue="J8:13", active_high=False)
##
##define the pins for R:GPIO17,G:GPIO18,B:GPIO27 
##
led = RGBLED(red=17, green=18, blue=27, active_high=False)



########################################################################################
##
## change duty cycle for the three specified pins to r_val,g_val,b_val
##
def setColor(r_val,g_val,b_val):

    ####################################################################################
    ##
    ## change pwmRed duty cycle to r_val
    ##
    led.red=r_val/100
    
    ####################################################################################
    ##
    ## change pwmGreen duty cycle to g_value
    ##
    led.green = g_val/100
    
    ####################################################################################
    ##
    ## change pwmBlue duty cycle to b_val
    ##
    led.blue = b_val/100

def loop():
    while True:
    
        ################################################################################
        ##
        ## get three different random numbers in the range 0-100
        ##
        r=random.randint(0,100)  
        g=random.randint(0,100)
        b=random.randint(0,100)
        
        ################################################################################
        ##
        ## set the random numbers as the duty cycle values
        ##
        setColor(r,g,b)
        
        ################################################################################
        ##
        ## print the current RGB values
        ##
        print ('r=%d, g=%d, b=%d ' %(r ,g, b))
        
        
        ################################################################################
        ##
        ## strategic delay for user perception (1 second)
        ##
        time.sleep(1)
        
def destroy():

    ####################################################################################
    ##
    ## closes the connection to the LED pin
    ##
    led.close()

########################################################################################
##
## start of program
##
if __name__ == '__main__':

    ####################################################################################
    ##
    ## display operational information
    ##
    print ('Program is starting ... ')
    
    ####################################################################################
    ##
    ## normal operations: run the loop
    ##
    try:
        loop()
        
    ####################################################################################
    ##
    ## catch any keyboard interrupt characters to signal program exit
    ##
    except KeyboardInterrupt:  # Press ctrl-c to end the program.
        destroy()
        print("Ending program")

File: code/C++/05.1.1_colorfulLED/colorfulLED.cc
//
// Filename:    colorfulLED.cc
// Description: Random color change on RGB LED
// Components:  RGB LED
//
////////////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////////////
//
// Pre-processor directives
//
#include <cstdint>
#include <cstdlib>
#include <wiringPi.h>
#include <softPwm.h>
#include <iostream>

////////////////////////////////////////////////////////////////////////////////////////
//
// Class for storing LED data
//
class LED
{
public:
    
    ////////////////////////////////////////////////////////////////////////////////////////
    //
    // Constructors
    //
    LED();
    LED(int32_t redval, int32_t grnval, int32_t bluval, int32_t redpin, int32_t grnpin, int32_t blupin);

    ////////////////////////////////////////////////////////////////////////////////////////
    //
    // Function prototypes
    //
    void LEDsetup    (void);
    void LEDsetcolor (int32_t redval, int32_t grnval, int32_t bluval);
    
    void setRed(int32_t value);
    void setGreen(int32_t value);
    void setBlue(int32_t value);
    
    int32_t getRed();
    int32_t getGreen();
    int32_t getBlue();

private:

    ////////////////////////////////////////////////////////////////////////////////////////
    //
    // Variables for storing the LEDs rgb values and pins
    //
    int32_t  red;
    int32_t  green;
    int32_t  blue;     
    
    int32_t  REDpin;//  0
    int32_t  GRNpin;//  1
    int32_t  BLUpin;//  2
};

int32_t  main (void)
{
    
    ////////////////////////////////////////////////////////////////////////////////////
    //
    // Instantiate the LED class
    //
    LED newLED;
    
    ////////////////////////////////////////////////////////////////////////////////////
    //
    // Bring wiringPi functionality online
    //
    if (wiringPiSetup () == -1)
    {
        std::cerr << "[ERROR] Could not initialize wiringPi library!\n";
        exit (1);
    }

    ////////////////////////////////////////////////////////////////////////////////////
    //
    // Configure the RGB LED for software PWM operations
    //
    newLED.LEDsetup ();

    std::cout << "Program is starting (CTRL-c to interrupt) ...\n";

    while (1)
    {
        ////////////////////////////////////////////////////////////////////////////////
        //
        // For each of red, green, blue: obtain a pseudo-random value between 0 and
        // 100 (the configured software PWM range established)
        //
        newLED.setRed(random () % 100);
        newLED.setGreen(random () % 100);
        newLED.setBlue(random () % 100);

        ////////////////////////////////////////////////////////////////////////////////
        //
        // Set the RGB LED to the random values
        //
        newLED.LEDsetcolor (newLED.getRed(), newLED.getGreen(), newLED.getBlue());

        ////////////////////////////////////////////////////////////////////////////////
        //
        // For informative feedback, display these values to the user
        //
        std::cout << "red: " << newLED.getRed() << ", green: " << newLED.getGreen() << ", blue: " << newLED.getBlue() << "\n";

        ////////////////////////////////////////////////////////////////////////////////
        //
        // Strategic delay to allow for better user perception (1000mS == 1 second)
        //
        delay (1000);
    }

    return (0);
}

////////////////////////////////////////////////////////////////////////////////////////
//
// LED::LED(): the parameterless constructor, initialize the member attributes to 
// default values
//
LED::LED()
{
    red = 0;
    green = 0;
    blue = 0;
    REDpin = 0;
    GRNpin = 1;
    BLUpin = 2;
}

////////////////////////////////////////////////////////////////////////////////////////
//
// LED::LED(redval, grnval, bluval, redpin, grnpin, blupin): the parametered constructor, 
// initialize the member attributes to provided values
//
LED::LED(int32_t redval, int32_t grnval, int32_t bluval, int32_t redpin, int32_t grnpin, int32_t blupin)
{
    red = redval;
    green = grnval;
    blue = bluval;
    REDpin = redpin;
    GRNpin = grnpin;
    BLUpin = blupin;
}

////////////////////////////////////////////////////////////////////////////////////////
//
// Configure software PWM on the designated RGB LED GPIO pins, establishing the
// operational range of 0-100.
//
void LED::LEDsetup (void)
{
    softPwmCreate (REDpin, 0, 100); // Create SoftPWM pin for red
    softPwmCreate (GRNpin, 0, 100); // Create SoftPWM pin for green
    softPwmCreate (BLUpin, 0, 100); // Create SoftPWM pin for blue
}

////////////////////////////////////////////////////////////////////////////////////////
//
// LEDsetcolor(): update the PWM values for the RGB LED.
//
void LED::LEDsetcolor (int32_t redval, int32_t grnval, int32_t bluval)
{
    softPwmWrite (REDpin, redval);    // Set the REDpin to redval
    softPwmWrite (GRNpin, grnval);  // Set the GRNpin to grnval
    softPwmWrite (BLUpin, bluval);  // Set the BLUpin to bluval
}

////////////////////////////////////////////////////////////////////////////////////////
//
// LED::setRed(): set the "red" attribute to the provided value
//
void LED::setRed(int32_t value)
{
    red = value;
}

////////////////////////////////////////////////////////////////////////////////////////
//
// LED::setGreen(): set the "green" attribute to the provided value
//
void LED::setGreen(int32_t value)
{
    green = value;
}

////////////////////////////////////////////////////////////////////////////////////////
//
// LED::setblue(): set the "blue" attribute to the provided value
//
void LED::setBlue(int32_t value)
{
    blue = value;
}

////////////////////////////////////////////////////////////////////////////////////////
//
// LED::getRed(): return the value of the "red" attribute
//
int32_t LED::getRed()
{
    return red;
}

////////////////////////////////////////////////////////////////////////////////////////
//
// LED::getGreen(): return the value of the "green" attribute
//
int32_t LED::getGreen()
{
    return green;
}

////////////////////////////////////////////////////////////////////////////////////////
//
// LED::getBlue(): return the value of the "blue" attribute
//
int32_t LED::getBlue()
{
    return blue;
}
